### PR TITLE
boot: bootutil: Simplify check for crypto backends to reduce complexity

### PR DESCRIPTION
**Pull request by sigvartmh**
Adding multiple crypto backends will grow quadraticly. This change will
ensure that the growth will be linear and generate less complexity.

Signed-off-by: Sigvart Hovland <sigvart.m@gmail.com>

### CODE REVIEW

**Comment by sigvartmh** [comment]
@mbolivar @utzig @nvlsianpu

File: boot/bootutil/include/bootutil/sha256.h [hunk 1]
```diff
@@ -31,13 +31,12 @@
...
-#endif
-
-#if !defined(MCUBOOT_USE_MBED_TLS) && !defined(MCUBOOT_USE_TINYCRYPT) && !defined(MCUBOOT_USE_CC310)
+#define NUM_CRYPTO_BACKENDS (defined(MCUBOOT_USE_MBED_TLS) +\
+			     defined(MCUBOOT_USE_TINYCRYPT) +\
+			     defined(MCUBOOT_USE_CC310))
+#if NUM_CRYPTO_BACKENDS > 1
+    #error "Cannot define pairs of CC310, MBED_TLS, and TINYCRYPT"
+#elif NUM_CRYPTO_BACKENDS == 0
     #error "One of CC310, MBED_TLS or TINYCRYPT must be defined"
```
Comment (utzig) [inline boot/bootutil/include/bootutil/sha256.h]: This message would also apply to the first test since it's "One of ..."

Comment (sigvartmh) [reply boot/bootutil/include/bootutil/sha256.h]: Fixed

Comment (sigvartmh) [reply boot/bootutil/include/bootutil/sha256.h]: Maybe I should rephrase it to "Only one"

**Comment by sigvartmh** [comment]
Don't know why the CI fails

**Comment by utzig** [comment]
@sigvartmh You are trying to close an `#if...` that does not exist. Your code should probably be:

```
#define NUM_CRYPTO_BACKENDS (defined(MCUBOOT_USE_MBED_TLS) +\
			     defined(MCUBOOT_USE_TINYCRYPT) +\
			     defined(MCUBOOT_USE_CC310))
#if NUM_CRYPTO_BACKENDS != 1
    #error "One crypto backend must be defined, either CC310, MBED_TLS or TINYCRYPT"
#endif
```

**Comment by sigvartmh** [comment]
> @sigvartmh You are trying to close an `#if...` that does not exist. Your code should probably be:

@utzig : Woops was a bit too quick with modifying it what I wanted to write was
 ```
#if(defined(MCUBOOT_USE_MBED_TLS) +\
			     defined(MCUBOOT_USE_TINYCRYPT) +\
			     defined(MCUBOOT_USE_CC310)) != 1
    #error "One crypto backend must be defined either CC310, MBED_TLS or TINYCRYPT"
#endif
```

**Comment by utzig** [comment]
@sigvartmh Yup, that looks OK as well, but we are not running an obfuscated code competition, so might as well make it clearer...

**Comment by sigvartmh** [comment]
@utzig Yup, it's either solve by comment or naming so did the documentation by name approach as suggested.
